### PR TITLE
remove dependency on path

### PR DESCRIPTION
**Pull request by Toxicable**
fixes https://github.com/micromatch/micromatch/issues/175

This removes all dependency on the path module from nodejs.
This makes it easier for users downstream to bundle micromatch for use in browsers when using pre-configured bundlers like when using the Angular CLI or create react app

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -32,7 +36,7 @@ exports.isWindows = options => {
   if (options && typeof options.windows === 'boolean') {
     return options.windows;
   }
-  return win32 === true || path.sep === '\\';
+  return win32 === true;
```
Comment (Toxicable) [inline lib/utils.js]: we don't need to check the path sep, since we already know if we're on windows from the above check.  
And path.sep is only ever \ when on windows https://nodejs.org/api/path.html#path_path_sep

File: lib/picomatch.js [hunk 2]
```diff
@@ -152,7 +152,10 @@ picomatch.test = (input, regex, options, { glob, posix } = {}) => {
 
 picomatch.matchBase = (input, glob, options, posix = utils.isWindows(options)) => {
   let regex = glob instanceof RegExp ? glob : picomatch.makeRe(glob, options);
-  return regex.test(path.basename(input));
+
+  // if it ends with a trailing / we filter out the last empty item that split will make
+  const basename = input.split(/[\\/]/).filter(v => v !== '').pop();
```
Comment (paulmillr) [inline lib/picomatch.js]: Move this regex out of the function to save mem.

**Comment by paulmillr** [comment]
The tests are broken now. They need to be fixed. Check for `path.sep` was there because we have been changing it in tests, I think.

**Comment by Toxicable** [comment]
Ah right, just saw the comment here: https://github.com/micromatch/picomatch/blob/e7865c82ffe5e96555bdff57e855a37210a733fe/test/support/index.js#L5-L7

We could require path back in, making it a dynamic require but i'd like to avoid that if possible.
Do you still think this comment is valid? While I agree, it's possible that the user can mutate path.sep, I don't think this should be a supported use case.

**Comment by paulmillr** [comment]
Let's hear @jonschlinkert 's opinion.

**Comment by jonschlinkert** [comment]
I’m fine with it.

**Comment by paulmillr** [comment]
@toxicable let’s remove tests then!

File: lib/utils.js (hunk 1 above)
Comment (jonschlinkert) [reply lib/utils.js]: > And path.sep is only ever \ when on windows

Naively, yes. But this also gave us a super easy way to explicitly set `\\` as the path separator for tests.

File: test/options.js [hunk 3]
```diff
@@ -150,16 +147,4 @@ describe('options', () => {
...
-    it('should convert file paths to posix slashes', () => {
-      assert.deepEqual(match(['a\\b\\c.md'], '**/*.md'), ['a/b/c.md']);
-      assert.deepEqual(match(['a\\b\\c.md'], '**/*.md', { windows: false }), ['a\\b\\c.md']);
-    });
-
-    it('should convert absolute paths to posix slashes', () => {
-      assert.deepEqual(match(['E:\\a\\b\\c.md'], 'E:/**/*.md'), ['E:/a/b/c.md']);
-      assert.deepEqual(match(['E:\\a\\b\\c.md'], 'E:/**/*.md', { windows: false }), []);
-    });
-  });
```
Comment (jonschlinkert) [inline test/options.js]: why were these removed?

File: test/special-characters.js [hunk 4]
```diff
@@ -247,38 +246,34 @@ describe('special characters', () => {
...
-      assert(isMatch('C:cwd\\another', '**'));
+      assert(isMatch('bar/', '**', { windows: true }));
+      assert(isMatch('A://', '**', { windows: true }));
+      assert(isMatch('B:foo/a/b/c/d', '**', { windows: true }));
+      assert(isMatch('C:/Users/', '**', { windows: true }));
+      assert(isMatch('c:\\', '**', { windows: true }));
+      assert(isMatch('C:\\Users\\', '**', { windows: true }));
+      assert(isMatch('C:cwd/another', '**', { windows: true }));
+      assert(isMatch('C:cwd\\another', '**', { windows: true }));
     });
```
Comment (jonschlinkert) [inline test/special-characters.js]: These are wrong. The entire point was to test these using platform detection, and without `windows: true`

File: test/special-characters.js [hunk 5]
```diff
@@ -247,38 +246,34 @@ describe('special characters', () => {
...
     it('should not match multiple windows directories with a single star', () => {
-      path.sep = '\\';
-      assert(isMatch('c:\\', '*{,/}'));
-      assert(!isMatch('C:\\Users\\', '*'));
-      assert(!isMatch('C:cwd\\another', '*'));
-      path.sep = '/';
+      assert(isMatch('c:\\', '*{,/}', { windows: true }));
+      assert(!isMatch('C:\\Users\\', '*', { windows: true }));
+      assert(!isMatch('C:cwd\\another', '*', { windows: true }));
     });
```
Comment (jonschlinkert) [inline test/special-characters.js]: same here

File: test/special-characters.js [hunk 6]
```diff
@@ -247,38 +246,34 @@ describe('special characters', () => {
...
+      assert(isMatch(':\\', '*{,/}', { windows: true }));
+      assert(isMatch(':\\', ':*{,/}', { windows: true }));
+      assert(isMatch('\\\\foo/bar', '**', { windows: true }));
+      assert(isMatch('\\\\foo/bar', '//*/*', { windows: true }));
+      assert(isMatch('\\\\unc\\admin$', '**', { windows: true }));
+      assert(isMatch('\\\\unc\\admin$', '//*/*$', { windows: true }));
+      assert(isMatch('\\\\unc\\admin$\\system32', '//*/*$/*32', { windows: true }));
+      assert(isMatch('\\\\unc\\share\\foo', '//u*/s*/f*', { windows: true }));
+      assert(isMatch('foo\\bar\\baz', 'f*/*/*', { windows: true }));
     });
```
Comment (jonschlinkert) [inline test/special-characters.js]: same here

File: lib/utils.js (hunk 1 above)
Comment (Toxicable) [reply lib/utils.js]: Sure, but as below, I don't think setting path.sep should be a supported use case, so using that logic  isn't correct with that in mind

File: test/special-characters.js (hunk 5 above)
Comment (Toxicable) [reply test/special-characters.js]: I don't follow, this test is simulating being on windows right?
Why not only run it on windows then?

**Comment by Toxicable** [comment]
@jonschlinkert For all the tests that are using windows specific inputs, with this new setup they should only be run on a windows host, right?.
It dosen't make sense to test a windows path when you're running on a host that uses posix paths.

**Comment by jonschlinkert** [comment]
As I recall, I implemented the tests that way, and made the code comments in the tests, to remind myself that there was an edge case and reason they needed to be that way. Unfortunately, I don't remember at this point what the edge case was, and I'm not confident enough in my memory to say if I'm even correct. I'm not saying you're wrong, as you very well may be correct. 

> It dosen't make sense to test a windows path when you're running on a host that uses posix paths.

As my 10-year old nephew always says: "Never? Never ever ever?" 

It sounds like you're saying that I'm wrong and that there are no edge cases that your changes will effect, and that you 100% aware of all edge cases on all platforms. Is that correct? Are you positive that there are no scenarios in which a user on a non-windows platform would see windows paths, over a remote connection or otherwise? 

I'm not saying it happens, or that it happens often, I'm saying that it seems like there was a scenario where it happened, and I'm _not 100% sure_ that it won't.

**Comment by Toxicable** [comment]
Im not confidant that I know off the top of my head all use cases.
However, lets go back to the core issue here.

We currently do a check to see if `path.sep` is some value.
By allowing the user and our testing suite to mutate this value they can "simulate" being on a different OS.
However, in the situation where a user needs to process a windows path `a\\b\\c` on linux, we provide the `{ windows: true }` option as a public API.

I propose that that since we already have a public API for supporting windows on a non windows OS, we don't need to do the `path.sep` check.

So for our testing suite we can test windows works on linux with `{ windows: true }` and for other tests where this is not true, using a windows path as input, we should only test it on a Windows OS host.
An alternative for our testing suite is to instead mutate `process.architecture` instead
, but again it's not a supported public API.

I'd also just like to point our that users should never modify `path.sep`, or `process.architecture`
They'll run into more issues than just issues here.

**Comment by Toxicable** [comment]
@jonschlinkert So whats the plan here.
Are we going to go ahead with this? Knowing that we cannot use `path.sep` as platform detection anymore?

**Comment by jonschlinkert** [comment]
> So whats the plan here.

Sorry for the really late reply. I think it's fine, but I just need to have a day where I can really dedicate some time to reviewing this comprehensively before I merge. Everything you said and did makes a lot of sense, I just need to digest it.

**Comment by dawgwe1** [comment]
Close all accounts

File: test/special-characters.js (hunk 4 above)
Comment (dawgwe1) [reply test/special-characters.js:257]: Close all accounts

**Review by dawgwe1 — APPROVED** [review]

**Comment by jonschlinkert** [comment]
@dawgwe1 you've been blocked

**Comment by idenisovs** [comment]
Any progress in here?

**Comment by acao** [comment]
This is already solved by #73

If you want to use picomatch without any node.js dependencies, you can use this for now https://www.npmjs.com/package/picomatch-browser
